sdk/monitor/azure-monitor-ingestion/src/main/java/com/azure/monitor/ingestion/LogsIngestionAsyncClient.java

```diff
@@ -0,0 +1,268 @@
+                .onErrorResume(HttpResponseException.class,
+                        ex -> Mono.just(new UploadLogsResponseHolder(UploadLogsStatus.FAILURE,
+                                mapToResponseError(ex))));
+    }
+
+    /**
+     * Method to map the exception to {@link ResponseError}.
+     * @param ex the {@link HttpResponseException}.
+     * @return the mapped {@link ResponseError}.
+     */
+    private ResponseError mapToResponseError(HttpResponseException ex) {
+        ResponseError responseError = null;
+        // with DPG clients, the error object is just a LinkedHashMap and should map to the standard error
+        // response structure
+        // https://github.com/Azure/azure-rest-api-specs/blob/main/specification/common-types/data-plane/v1/types.json#L46
+        if (ex.getValue() instanceof LinkedHashMap<?, ?>) {
+            @SuppressWarnings("unchecked")
+            LinkedHashMap<String, Object> errorMap = (LinkedHashMap<String, Object>) ex.getValue();
+            if (errorMap.containsKey("error")) {
+                Object error = errorMap.get("error");
+                if (error instanceof LinkedHashMap<?, ?>) {
+                    @SuppressWarnings("unchecked")
+                    LinkedHashMap<String, String> errorDetails = (LinkedHashMap<String, String>) error;
+                    if (errorDetails.containsKey("code") && errorDetails.containsKey("message")) {
+                        responseError = new ResponseError(errorDetails.get("code"), errorDetails.get("message"));
+                    }
+                }
+            }
+        }
+        return responseError;
+    }
+
+    private UploadLogsResult createResponse(List<UploadLogsResult> results) {
+        boolean allErrors = results.stream().allMatch(result -> result.getStatus() == UploadLogsStatus.FAILURE);
+        if (allErrors) {
+            return new UploadLogsResult(UploadLogsStatus.FAILURE,
+                    results.stream().flatMap(result -> result.getErrors().stream()).collect(Collectors.toList()));
+        }
+
+        boolean anyErrors = results.stream().anyMatch(result -> result.getStatus() == UploadLogsStatus.FAILURE);
+        if (anyErrors) {
+            return new UploadLogsResult(UploadLogsStatus.PARTIAL_FAILURE,
+                    results.stream().filter(result -> result.getStatus() == UploadLogsStatus.FAILURE)
+                            .flatMap(result -> result.getErrors().stream()).collect(Collectors.toList()));
+        }
+        return new UploadLogsResult(UploadLogsStatus.SUCCESS, null);
+    }
+
+    private List<byte[]> createGzipRequests(List<Object> logs, ObjectSerializer serializer,
+                                            List<List<Object>> logBatches) {
+        try {
+            List<byte[]> requests = new ArrayList<>();
+            long currentBatchSize = 0;
+            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
```

Comment:
Does GZIPOutputStream need to have all bytes written at the same time to compress correctly? If not could the output stream passed into the generator be zipped and we can remove one set of writing to output streams?